### PR TITLE
Core - Implemented Tagged Data Payload 

### LINKED ISSUE #290: Core - Implement tagged data
## Description

Add support for [tagged data](https://github.com/Wollac/protocol-rfcs/blob/tagged-data/tips/TIP-0023/tip-0023.md) payload.

### PR DESCRIPTION
**Pull request by matjazv**
# Description of change

Fixes #290 

## Type of change

- Enhancement (a non-breaking change which adds functionality)
- Breaking change (fix or feature that would cause existing functionality to not work as expected)

## How the change has been tested

Unit tests implemented in `core/test_tagged_data.c`

## Change checklist

Add an `x` to the boxes that are relevant to your changes, and delete any items that are not.

- [x] My code follows the contribution guidelines for this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests using CTest that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: src/core/models/payloads/tagged_data.c [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+
+void tagged_data_free(tagged_data_t *tagged_data) {
+  if (tagged_data) {
+    byte_buf_free(tagged_data->tag);
+    byte_buf_free(tagged_data->data);
+    free(tagged_data);
+  }
+}
+
+tagged_data_t *tagged_data_create(char const *tag, byte_t data[], uint32_t data_len) {
```
Comment (oopsmonk) [inline src/core/models/payloads/tagged_data.c]: `char const *tag` indcates it's a pointer to a single character but the tag is an array, suggest change to
```suggestion
tagged_data_t *tagged_data_create(char const tag[], byte_t data[], uint32_t data_len) {
```

File: src/core/models/payloads/tagged_data.c [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+
+  // add tag string
+  if (strlen(tag) > 0) {
+    byte_t tag_bin[TAGGED_DATA_TAG_MAX_LENGTH_BYTES] = {0};
+    if (string2hex(tag, tag_bin, tag_hex_len) != 0) {
+      printf("[%s:%d] can not convert string into a hex\n", __func__, __LINE__);
+      tagged_data_free(tagged_data);
+      return NULL;
+    }
+    if (!byte_buf_set(tagged_data->tag, tag_bin, tag_hex_len)) {
```
Comment (oopsmonk) [inline src/core/models/payloads/tagged_data.c]: tag is optional, maybe set `tagged_data->tag` to NULL in `tagged_data_new()` and use `byte_buf_new_with_data` here, for a better performance?

File: src/core/models/payloads/tagged_data.c [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+      printf("[%s:%d] adding tag to a tagged data failed\n", __func__, __LINE__);
+      tagged_data_free(tagged_data);
+      return NULL;
+    }
+  }
+
+  // add binary data
+  if (data_len > 0) {
```
Comment (oopsmonk) [inline src/core/models/payloads/tagged_data.c:64]: data is mandatory, it's a redundant check, `data_len` has been checked at line 36.
`byte_buf_new_with_data` can be used here too.
```suggestion

```

**Review by oopsmonk — CHANGES_REQUESTED** [review]

File: src/core/models/payloads/tagged_data.c (hunk 1 above)
Comment (matjazv) [reply src/core/models/payloads/tagged_data.c]: Done, thanks.

File: src/core/models/payloads/tagged_data.c (hunk 2 above)
Comment (matjazv) [reply src/core/models/payloads/tagged_data.c]: Fixed, thanks.

File: src/core/models/payloads/tagged_data.c (hunk 3 above)
Comment (oopsmonk) [reply src/core/models/payloads/tagged_data.c:64]: Please ignore, data can be empty in some cases.

File: src/core/models/payloads/tagged_data.c [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+}
+
+void tagged_data_free(tagged_data_t *tagged_data) {
+  if (tagged_data) {
+    if (tagged_data->tag) {
+      byte_buf_free(tagged_data->tag);
+    }
+    if (tagged_data->data) {
+      byte_buf_free(tagged_data->data);
+    }
```
Comment (oopsmonk) [inline src/core/models/payloads/tagged_data.c]: the point is checked in `byte_buf_free` too.
```suggestion
    byte_buf_free(tagged_data->tag);
    byte_buf_free(tagged_data->data);
```

**Review by oopsmonk — APPROVED** [review]

File: src/core/models/payloads/tagged_data.c (hunk 4 above)
Comment (matjazv) [reply src/core/models/payloads/tagged_data.c]: Thanks. It's fixed.

**Pull request merged**